### PR TITLE
Added multithreading to BinaryArithmetic to increase speed. 

### PR DESCRIPTION
**Pull request by KingAkeem**
This PR references issue #21 I've added multithreading code for concurrently running four `strip_leading_zeroes` functions. This has reduced the test time of the BinaryArithemetic tests by 3x as much. I've attached some photos of comparison below:
Before multithreading:
screen shot 2018-05-10 at 6 20 28 am
After multithreading:
screen shot 2018-05-10 at 7 38 33 am

### CODE REVIEW

**Comment by faheel** [comment]
This is great! Can you also take a look at other places where multithreading can be used to speed up things? For example, calculating `prod_high`, `prod_mid` and `prod_low` on different threads may lead to a significant speedup.

**Comment by KingAkeem** [comment]
Thank you! I definitely was planning on it. I just wanted to get a review on what I had already done to show the improvement and let you know that I was working on it.

File: include/operators/binary_arithmetic.hpp [hunk 1]
```diff
@@ -180,10 +182,15 @@ BigInt BigInt::operator*(const BigInt& num) const {
...
         num2_low = smaller.substr(half_length);
 
-        strip_leading_zeroes(num1_high.value);
-        strip_leading_zeroes(num1_low.value);
-        strip_leading_zeroes(num2_high.value);
-        strip_leading_zeroes(num2_low.value);
+        std::thread t1(strip_leading_zeroes, std::ref(num1_high.value));
+        std::thread t2(strip_leading_zeroes, std::ref(num1_low.value));
+        std::thread t3(strip_leading_zeroes, std::ref(num2_high.value));
+        std::thread t4(strip_leading_zeroes, std::ref(num2_low.value));
```
Comment (faheel) [inline include/operators/binary_arithmetic.hpp]: The threads `t1`, `t2`, ... aren't very descriptive and should be renamed to `thread1`, `thread2`, ...

Comment (KingAkeem) [reply include/operators/binary_arithmetic.hpp]: Added the fix :)

**Comment by faheel** [comment]
@KingAkeem I've updated the testing and build mechanism recently, so can you merge the recent changes to your branch and push it again to that we can if the build still fails (only 1 build, that uses GCC 8 and does coverage analysis should fail, the rest should pass).

**Comment by faheel** [comment]
@KingAkeem Please merge with latest upstream (for the reason mentioned above)

**Comment by KingAkeem** [comment]
@faheel I've merged the recent changes and pushed the branch. Sorry about taking so long, I just saw your comment.

**Comment by faheel** [comment]
Looks like on Linux we'll need to add `-pthread` when compiling to link the pthread library. It's not that big of a deal, but then the compilation step will require special attention from the user which impacts the main features of this library, which are simplicity and ease of use. So I'll need to think about this and see what's the best thing to do here.

What are your thoughts on this @KingAkeem?

**Comment by xorz57** [comment]
@faheel It's impossible to use STL Threads without linking with -pthread.
I don't think it's hard for people to read the documentation and see that they have to link with -pthread.

**Pull request closed without merging**